Add support for dictionary with torch object keys.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103158

Fixes: #101979

This PR adds support for dictionaries with torch object as keys in dynamo.

The main problem was that, for example, the source built for `d[torch.float]` (`d` being a
dictionary) was `ODictGetItemSource(GlobalSource('d'), index=torch.float)`. When
`Source.name` method was called, we got `odict_getitem(G['d'], torch.float)`. Evaluating
that string raised an error, since `torch` was only available in the global dictionary `G`
as `G["torch"]`.

Instead, this PR builds the source:
`ODictGetItemSource(GlobalSource('d'), index=AttrSource(GlobalSource('torch'), 'float'))`.
The to-be-evaluated string is correctly generated as:
`odict_getitem(G['d'], G['torch'].float)`.

Here's a minimal example that reproduces the error, before this PR:

```python
import torch

d = {
    torch.float16: torch.float32,
}

@torch.compile
def f():
    return torch.randn(3, dtype=d[torch.float16])

f()
```

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @aakhundov

ysiraichi committed:
Add support for dictionary with torch object keys.

Fixes: #101979

This PR adds support for dictionaries with torch object as keys in dynamo.

The main problem was that, for example, the source built for `d[torch.float]` (`d` being a
dictionary) was `ODictGetItemSource(GlobalSource('d'), index=torch.float)`. When
`Source.name` method was called, we got `odict_getitem(G['d'], torch.float)`. Evaluating
that string raised an error, since `torch` was only available in the global dictionary `G`
as `G["torch"]`.

Instead, this PR builds the source:
`ODictGetItemSource(GlobalSource('d'), index=AttrSource(GlobalSource('torch'), 'float'))`.
The to-be-evaluated string is correctly generated as:
`odict_getitem(G['d'], G['torch'].float)`.

Here's a minimal example that reproduces the error, before this PR:

```python
import torch

d = {
    torch.float16: torch.float32,
}

@torch.compile
def f():
    return torch.randn(3, dtype=d[torch.float16])

f()
```

[ghstack-poisoned]

ysiraichi committed:
Fix reconstruct method. on "Add support for dictionary with torch object keys."


Fixes: #101979

This PR adds support for dictionaries with torch object as keys in dynamo.

The main problem was that, for example, the source built for `d[torch.float]` (`d` being a
dictionary) was `ODictGetItemSource(GlobalSource('d'), index=torch.float)`. When
`Source.name` method was called, we got `odict_getitem(G['d'], torch.float)`. Evaluating
that string raised an error, since `torch` was only available in the global dictionary `G`
as `G["torch"]`.

Instead, this PR builds the source:
`ODictGetItemSource(GlobalSource('d'), index=AttrSource(GlobalSource('torch'), 'float'))`.
The to-be-evaluated string is correctly generated as:
`odict_getitem(G['d'], G['torch'].float)`.

Here's a minimal example that reproduces the error, before this PR:

```python
import torch

d = {
    torch.float16: torch.float32,
}

torch.compile
def f():
    return torch.randn(3, dtype=d[torch.float16])

f()
```

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy

[ghstack-poisoned]

ysiraichi committed:
Fix lint. on "Add support for dictionary with torch object keys."


Fixes: #101979

This PR adds support for dictionaries with torch object as keys in dynamo.

The main problem was that, for example, the source built for `d[torch.float]` (`d` being a
dictionary) was `ODictGetItemSource(GlobalSource('d'), index=torch.float)`. When
`Source.name` method was called, we got `odict_getitem(G['d'], torch.float)`. Evaluating
that string raised an error, since `torch` was only available in the global dictionary `G`
as `G["torch"]`.

Instead, this PR builds the source:
`ODictGetItemSource(GlobalSource('d'), index=AttrSource(GlobalSource('torch'), 'float'))`.
The to-be-evaluated string is correctly generated as:
`odict_getitem(G['d'], G['torch'].float)`.

Here's a minimal example that reproduces the error, before this PR:

```python
import torch

d = {
    torch.float16: torch.float32,
}

torch.compile
def f():
    return torch.randn(3, dtype=d[torch.float16])

f()
```

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy

[ghstack-poisoned]

mlazos commented on torch/_dynamo/source.py:
```diff
@@ -399,10 +399,15 @@ def __post_init__(self):
         assert self.base is not None
 
     def reconstruct(self, codegen):
```
This is nearly identical to the standard `GetItemSource` above, can you abstract this out to a common place?

ysiraichi replied:
As far as I understand, the identical parts of the 2 are:
- Reconstruct the `base`
- Check if the `index` is an instance of `Source`
- Maybe (check the differences) load the `index` if it is not 

The differences are:
- `GetItemSource` still has to check if `index` is a slice
- `GetItemSource` creates a `BINARY_SUBSCR` instruction
- `ODictGetItemSource` has to load the `getitem` function
- `ODictGetItemSource` has to call this function (another bytecode)

Do you think that I should do it, regardless of the differences?

mlazos replied:
If you abstract out a bytecode_suffix method that is overridden in each subclass, and say index_is_slice is always false for OdictGetItem, I think the differences resolve, right?

ysiraichi replied:
Yes. You are totally right. Missed that completely.

mlazos commented on torch/_dynamo/variables/user_defined.py:
```diff
@@ -502,14 +502,13 @@ def call_hasattr(self, tx, name: str) -> "VariableTracker":
     def odict_getitem(self, tx, key):
         from .builder import VariableBuilder
 
+        index = key.source if key.source is not None else key.as_python_constant()
```
You should check directly if this is a supported type vs checking if the source is not `None`. A lot of types will meet this constraint

ysiraichi replied:
Should I do it like in [this comment](https://github.com/pytorch/pytorch/issues/101979#issuecomment-1574438751)?

mlazos replied:
So I'm thinking we should have an `is_valid_keys` function which decides whether a type is permitted or not, this already exists for `ConstDictVariable`. Can you augment this to check if it is a valid `TorchVariable` representing a dtype and make sure both `ConstDictVariable` and ODicts use that function to check for valid keys?

ysiraichi committed:
Refactor + add tests. on "Add support for dictionary with torch object keys."


Fixes: #101979

This PR adds support for dictionaries with torch object as keys in dynamo.

The main problem was that, for example, the source built for `d[torch.float]` (`d` being a
dictionary) was `ODictGetItemSource(GlobalSource('d'), index=torch.float)`. When
`Source.name` method was called, we got `odict_getitem(G['d'], torch.float)`. Evaluating
that string raised an error, since `torch` was only available in the global dictionary `G`
as `G["torch"]`.

Instead, this PR builds the source:
`ODictGetItemSource(GlobalSource('d'), index=AttrSource(GlobalSource('torch'), 'float'))`.
The to-be-evaluated string is correctly generated as:
`odict_getitem(G['d'], G['torch'].float)`.

Here's a minimal example that reproduces the error, before this PR:

```python
import torch

d = {
    torch.float16: torch.float32,
}

torch.compile
def f():
    return torch.randn(3, dtype=d[torch.float16])

f()
```

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy

[ghstack-poisoned]

ysiraichi commented on torch/_dynamo/variables/user_defined.py:
```diff
@@ -502,14 +503,18 @@ def call_hasattr(self, tx, name: str) -> "VariableTracker":
     def odict_getitem(self, tx, key):
         from .builder import VariableBuilder
 
+        index = (
+            key.source
+            if ConstDictVariable.is_valid_key(key) and key.source is not None
+            else key.as_python_constant()
+        )
```
Does this look better?

mlazos replied:
Yup thank you!

ysiraichi committed:
Fix lint. on "Add support for dictionary with torch object keys."


Fixes: #101979

This PR adds support for dictionaries with torch object as keys in dynamo.

The main problem was that, for example, the source built for `d[torch.float]` (`d` being a
dictionary) was `ODictGetItemSource(GlobalSource('d'), index=torch.float)`. When
`Source.name` method was called, we got `odict_getitem(G['d'], torch.float)`. Evaluating
that string raised an error, since `torch` was only available in the global dictionary `G`
as `G["torch"]`.

Instead, this PR builds the source:
`ODictGetItemSource(GlobalSource('d'), index=AttrSource(GlobalSource('torch'), 'float'))`.
The to-be-evaluated string is correctly generated as:
`odict_getitem(G['d'], G['torch'].float)`.

Here's a minimal example that reproduces the error, before this PR:

```python
import torch

d = {
    torch.float16: torch.float32,
}

torch.compile
def f():
    return torch.randn(3, dtype=d[torch.float16])

f()
```

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy aakhundov

[ghstack-poisoned]

ysiraichi committed:
Rebased. on "Add support for dictionary with torch object keys."


Fixes: #101979

This PR adds support for dictionaries with torch object as keys in dynamo.

The main problem was that, for example, the source built for `d[torch.float]` (`d` being a
dictionary) was `ODictGetItemSource(GlobalSource('d'), index=torch.float)`. When
`Source.name` method was called, we got `odict_getitem(G['d'], torch.float)`. Evaluating
that string raised an error, since `torch` was only available in the global dictionary `G`
as `G["torch"]`.

Instead, this PR builds the source:
`ODictGetItemSource(GlobalSource('d'), index=AttrSource(GlobalSource('torch'), 'float'))`.
The to-be-evaluated string is correctly generated as:
`odict_getitem(G['d'], G['torch'].float)`.

Here's a minimal example that reproduces the error, before this PR:

```python
import torch

d = {
    torch.float16: torch.float32,
}

torch.compile
def f():
    return torch.randn(3, dtype=d[torch.float16])

f()
```

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy aakhundov

[ghstack-poisoned]

ysiraichi commented:
@pytorchbot merge

ysiraichi commented:
@pytorchbot merge
